apps/dashboard/pkg/migration/conversion/v1beta1_to_v2alpha1.go

```diff
@@ -0,0 +1,2331 @@
+	x, y, width, height := int64(0), int64(0), int64(6), int64(3)
+
+	if gridPos, ok := panelMap["gridPos"].(map[string]interface{}); ok {
+		x = int64(getIntField(gridPos, "x", 0))
+		y = int64(getIntField(gridPos, "y", 0))
+		width = int64(getIntField(gridPos, "w", 6))
+		height = int64(getIntField(gridPos, "h", 3))
+	}
+
+	// Apply frontend-style grid position calculations
+	// Frontend recalculates positions based on row structure
+	if yOverride != nil {
+		y = *yOverride
+	}
+
+	item := dashv2alpha1.DashboardGridLayoutItemKind{
+		Kind: "GridLayoutItem",
+		Spec: dashv2alpha1.DashboardGridLayoutItemSpec{
+			X:      x,
+			Y:      y,
+			Width:  width,
+			Height: height,
+			Element: dashv2alpha1.DashboardElementReference{
+				Kind: "ElementReference",
+				Name: elementName,
+			},
+		},
+	}
+
+	// Handle repeat options
+	if repeat := schemaversion.GetStringValue(panelMap, "repeat"); repeat != "" {
+		repeatOptions := &dashv2alpha1.DashboardRepeatOptions{
+			Mode:  "variable",
+			Value: repeat,
+		}
+
+		if repeatDirection := schemaversion.GetStringValue(panelMap, "repeatDirection"); repeatDirection != "" {
+			switch repeatDirection {
+			case "h":
+				direction := dashv2alpha1.DashboardRepeatOptionsDirectionH
+				repeatOptions.Direction = &direction
+			case "v":
+				direction := dashv2alpha1.DashboardRepeatOptionsDirectionV
+				repeatOptions.Direction = &direction
+			}
+		}
+
+		if maxPerRow := getIntField(panelMap, "maxPerRow", 0); maxPerRow > 0 {
+			maxPerRowInt64 := int64(maxPerRow)
+			repeatOptions.MaxPerRow = &maxPerRowInt64
+		}
+
+		item.Spec.Repeat = repeatOptions
+	}
+
+	return item
+}
+
+func buildRowKind(rowPanelMap map[string]interface{}, elements []dashv2alpha1.DashboardGridLayoutItemKind) *dashv2alpha1.DashboardRowsLayoutRowKind {
+	collapsed := getBoolField(rowPanelMap, "collapsed", false)
+	title := schemaversion.GetStringValue(rowPanelMap, "title")
+
+	row := &dashv2alpha1.DashboardRowsLayoutRowKind{
+		Kind: "RowsLayoutRow",
+		Spec: dashv2alpha1.DashboardRowsLayoutRowSpec{
+			Collapse: &collapsed,
+			Title:    &title,
+			Layout: dashv2alpha1.DashboardGridLayoutKindOrAutoGridLayoutKindOrTabsLayoutKindOrRowsLayoutKind{
+				GridLayoutKind: &dashv2alpha1.DashboardGridLayoutKind{
+					Kind: "GridLayout",
+					Spec: dashv2alpha1.DashboardGridLayoutSpec{
+						Items: elements,
+					},
+				},
+			},
+		},
+	}
+
+	// Handle repeat options for rows
+	if repeat := schemaversion.GetStringValue(rowPanelMap, "repeat"); repeat != "" {
+		row.Spec.Repeat = &dashv2alpha1.DashboardRowRepeatOptions{
+			Mode:  "variable",
+			Value: repeat,
+		}
+	}
+
+	return row
+}
+
+func yOffsetInRows(panelMap map[string]interface{}, rowY int64) int64 {
+	if gridPos, ok := panelMap["gridPos"].(map[string]interface{}); ok {
+		panelY := int64(getIntField(gridPos, "y", 0))
+		return panelY - rowY - GRID_ROW_HEIGHT
+	}
+	return 0
+}
+
+const LEGACY_STRING_VALUE_KEY = "__legacyStringValue"
```

Comment:
could you add a comment on what this is used for?

Reply:
Done!